File: app/src/main/java/net/ginteam/carmen/manager/ApiManager.java
package net.ginteam.carmen.manager;

import android.util.Log;

import com.google.android.gms.maps.model.LatLng;

import net.ginteam.carmen.BuildConfig;
import net.ginteam.carmen.network.api.ApiLinks;
import net.ginteam.carmen.utils.ValidationUtils;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by Eugene on 12/23/16.
 */

public class ApiManager {

    private final int REQUEST_TIMEOUT_SECONDS = 15;

    private static ApiManager sInstance;

    private Retrofit mRetrofit;

    private ApiManager() {
        OkHttpClient httpClient = new OkHttpClient
                .Builder()
                .addInterceptor(new AuthorizationTokenInterceptor())
                .connectTimeout(REQUEST_TIMEOUT_SECONDS, TimeUnit.SECONDS)
                .build();

        mRetrofit = new Retrofit.Builder()
                .baseUrl(BuildConfig.BASE_API_URL)
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .addConverterFactory(GsonConverterFactory.create())
                .client(httpClient)
                .build();
    }

    public static ApiManager getInstance() {
        if (sInstance == null) {
            sInstance = new ApiManager();
        }
        return sInstance;
    }

    public <T> T getService(Class<T> serviceClass) {
        return mRetrofit.create(serviceClass);
    }

    private class AuthorizationTokenInterceptor implements Interceptor {

        @Override
        public Response intercept(Chain chain) throws IOException {
            Request apiRequest = chain.request();

            Log.d("HttpInterceptor", "Request URL: " + apiRequest.url().toString());

            Request.Builder builder = apiRequest.newBuilder();
            String authToken = PreferencesManager.getInstance().getUserToken();
            if (!authToken.isEmpty()) {
                Log.d("HttpInterceptor", "Auth token: " + authToken);

                builder.addHeader(ApiLinks.AUTH.AUTH_HEADER, "Bearer " + authToken);
            }

            LatLng userLocation = PreferencesManager.getInstance().getUserLocation();
            ValidationUtils.Pair<String, String> pair = ValidationUtils.validateUserLocation(userLocation);
            builder.addHeader(ApiLinks.CATALOG.LAT, pair.first).addHeader(ApiLinks.CATALOG.LNG, pair.second);

            apiRequest = builder.build();
            Log.d("HttpInterceptor", "Request headers: " + apiRequest.headers().toString());

            return chain.proceed(apiRequest);
        }

    }

}

File: app/src/main/kotlin/net/ginteam/carmen/kotlin/manager/ApiManager.kt
package net.ginteam.carmen.kotlin.manager

import android.util.Log
import net.ginteam.carmen.BuildConfig
import net.ginteam.carmen.kotlin.api.ApiSettings
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

/**
 * Created by eugene_shcherbinock on 2/10/17.
 */
object ApiManager {

    private const val CONNECTION_TIMEOUT: Long = 5

    val retrofit: Retrofit

    init {
        retrofit = Retrofit.Builder()
                .baseUrl(BuildConfig.BASE_API_URL)
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .addConverterFactory(GsonConverterFactory.create())
                .client(getHttpClient())
                .build()
    }

    private fun getHttpClient(): OkHttpClient {
        return OkHttpClient.Builder()
                .connectTimeout(CONNECTION_TIMEOUT, TimeUnit.SECONDS)
                .addInterceptor { chain ->
                    val userAccessToken = getUserAccessToken()
                    val locationHeader = getLocationHeader()

                    val originalRequest = chain.request()
                    val newRequest = originalRequest
                            .newBuilder()
                            .addHeader(ApiSettings.Auth.Params.AUTH_HEADER, userAccessToken)
                            .addHeader(ApiSettings.Catalog.Params.LATITUDE, locationHeader.first)
                            .addHeader(ApiSettings.Catalog.Params.LONGITUDE, locationHeader.second)
                            .build()

                    Log.d("ApiManager.kt", "Request URL: ${newRequest.url()}")
                    Log.d("ApiManager.kt", "Request headers: ${newRequest.headers()}")

                    chain.proceed(newRequest)
                }
                .build()
    }

    private fun getLocationHeader(): Pair <String, String> {
        val preferences: PreferencesManager = SharedPreferencesManager
        val location = preferences.userLocation
        if (location == null) {
            return Pair("0", "0")
        } else {
            return Pair(location.latitude.toString(), location.longitude.toString())
        }
    }

    private fun getUserAccessToken(): String {
        val preferences: PreferencesManager = SharedPreferencesManager
        return "Bearer ${preferences.userAccessToken}"
    }

}